### PR TITLE
Update index.d.ts

### PR DESCRIPTION
**Pull request by disjukr**
## Description

add id property to typedef

## Related Issues

https://github.com/timhagn/gatsby-background-image/issues/21
https://github.com/timhagn/gatsby-background-image/pull/23

### CODE REVIEW

**Comment by timhagn** [comment]
Oh, and many thanks for looking into this at all and opening a PR! Greatly appreciated : )!

**Comment by disjukr** [comment]
sorry i'm not good at english. but i just want to resolve this type error.

스크린샷 2019-09-18 오전 11 05 00

**Comment by timhagn** [comment]
\*hmmz\* Would have to look deeper into TypeScript, but we can't simply add all and every DOM-Property to `gbi`, methinks...
Just asked the nice folks on Twitter about it : ).

**Comment by cometkim** [comment]
It seems should to have definitely. And could be changed smarter way like: 
```diff
-interface BackgroundImageProps {
+interface BackgroundImageProps<T extends keyof IntrinsicElements ='div'> extends HTMLAttributes<T> {
...
-  Tag?: string;
+ Tag?: T;

```

**Comment by timhagn** [comment]
Hi @cometkim,

thanks for chiming in : )! Would you elaborate a little on "It seems should to have definitely.", cause I couldn't get the meaning by your code example, sorry : /. Tag can be a wide spectrum more than a `div`, or how is `T extends keyof JSX.IntrinsicElements` used in TS?

Thanks in advance,

best,

Tim

**Comment by cometkim** [comment]
Hey @timhagn Sure, let me explain in more detail.

TL;DR final code would be

```diff
+ type IntrinsicTags = keyof JSX.IntrinsicElements;
+ type InferExtraProps<T extends IntrinsicTags | void> = T extends infer U
+   ? U extends IntrinsicTags
+   ? U extends 'div'
+   ? { Tag?: 'div' } & JSX.IntrinsicElements['div']
+   : { Tag?: U } & JSX.IntrinsicElements[U]
+   : { Tag?: 'div' } & JSX.IntrinsicElements['div']
+   : { Tag?: 'div' } & JSX.IntrinsicElements['div']
+
+ interface IBackgroundImageProps {
    resolutions?: IFixedObject | (IFixedObject | string)[],
    sizes?: IFluidObject | (IFluidObject | string)[],
    fixed?: IFixedObject | (IFixedObject | string)[],
    fluid?: IFluidObject | (IFluidObject | string)[],
    fadeIn?: string | boolean,
    durationFadeIn?: number,
    title?: string,
    alt?: string
-   className?: string | object, // Support Glamor's css prop?.
    critical?: boolean,
    crossOrigin?: string | boolean,
-   style?: object | [], // Using PropTypes from RN. (not 100% sure of array type here)
    backgroundColor?: string | boolean,
    onLoad?: () => void
    onStartLoad?: (param: { wasCached: boolean }) => void
    onError?: (event: any) => void
-   Tag?: string,
    classId?: string,
    preserveStackingContext?: boolean,
  }

-  export default class BackgroundImage extends React.Component<IBackgroundProps>
+  export default class BackgroundImage<T extends IntrinsicTags> extends React.Component<InferExtraProps<T> & IBackgroundImageProps> {
  }
```

The original definition of BackgroundImage was just extends type of `React.Component` which is not allow to forward all global properties implicitly.

So we got type errors in strict mode, when using `key`, `id`, etc with BackgroundImage. this is why @disjukr got error above.

JSX.IntrinsicElements is helper type inside of `@types/react` provides every single detail type of HTML/SVG elements.

image

We can intersect it with original props for implicit extra props. And to be more precise, we can use type arguments witch would be inferred by TypeScript uses `Tag` prop.

For example, with new type definition.
```tsx
const A = <BackgroundImage Tag="div" width={30} /> // Type error, because the `width` is unknown property for <div> element.
const B = <BackgroundImage Tag="svg" width={30} /> // OK
```

This way is much helpful but also strict.

**Comment by timhagn** [comment]
Hey @cometkim,

wow, thanks a lot for the explanation, now I get TypeScript a little more : )! Might be a stupid follow-up question, but why does the `div` get a special handling with `JSX.IntrinsicElements['div']`?
Everything else is clear to me, now.

@disjukr Would you care to implement this, or would give @cometkim access rights so he'd be able to (if you want, @cometkim ; ).

Let's get this merged, soon!

Best,

Tim.

**Comment by cometkim** [comment]
>  but why does the div get a special handling with JSX.IntrinsicElements['div']

This is a hack :stuck_out_tongue_winking_eye: 

The default value for the GBI's `Tag` prop is `'div'`, also type arguments have 'div' as the default, is natural. And we should specify the type when we want to use a different `Tag` this case.

for exam:
```ts
export default class BackgroundImage<T extends IntrinsicTags = 'div'> extends React.Component<InferExtraProps<T> & IBackgroundImageProps> {}
// ^ T is always infered as 'div' here.
```
```tsx
<BackgroundImage Tag='section'/> // Type error, 'div' is incompatible with 'section'

<BackgroundImage<'section'> Tag='section' /> // correct.
```

But this is verbose to use. (double-typing)

I've found that the `infer` keyword attempts to infer more deeply when it is obscured by omitting some context (like `T='div'`). So, It's a little bit complicated to define the type that can omit Tag when only it is 'div', but it is possible as above.

**Comment by cometkim** [comment]
> Any good info about TypeScript out there, or where did you two learn it?

Actually switching to TS was not difficult for me, because I usually used Flow before.

@piotrwitek's [utility-types](https://github.com/piotrwitek/utility-types) is one of the good resources to learn advanced types of TS.

For the React, `@types/react` already provides types for most use cases so doesn't need complex types like this for most cases. You could easily make generic/type-friendly interfaces if you write them in TypeScript from scratch.

**Comment by timhagn** [comment]
Hi @cometkim,

thanks a lot for the write up, the commits and the recommendation : )!
Gonna try merging it on Friday at the latest, "have to be" in office from Monday through Thursday at the moment, so time is sparse (Homeoffice is in the making ; ).
Already worked with Flow a little (in React Native), so I'm definitely given Typescript a (second) chance % ).

Thanks an awefull lot, again!

Best,

Tim.

**Comment by timhagn** [comment]
Hi @cometkim & @disjukr,

thanks a lot to you both for the time invested to this great contribution! And thanks to @cometkim to take the time to bring me up to speed a little on TypeScript!

Merged and just published `v0.8.13` with it : ).

Best,

Tim.
